### PR TITLE
[Kernel] `Snapshot::publish` API with UC E2E test (using in-memory catalog client)

### PR DESCRIPTION
**Pull request by scottsand-db**
## 🥞 Stacked PR
Use this [link](https://github.com/delta-io/delta/pull/5332/files) to review incremental changes.
- [**stack/kernel_snapshot_publish**](https://github.com/delta-io/delta/pull/5332) [[Files changed](https://github.com/delta-io/delta/pull/5332/files)]

---------
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [X] Kernel
- [ ] Other (fill in here)

## Description

This PR adds the `Snapshot::publish` API. For some context, let's compare this to the commit API / flow:

- Commit: Connector --> (Kernel) `txn.commit(...)` --> (Committer) `committer.commit(...)`
- Publish: Connector --> (Kernel) `snapshot.publish(...)` --> (Committer) `committer.publish(...)`

In the future, this `publish` API should return the updated and published Snapshot.

## How was this patch tested?

New UC E2E tests (using in-memory catalog client).

## Does this PR introduce _any_ user-facing changes?

New `Snapshot::publish` API.

### CODE REVIEW

File: unity/src/test/scala/io/delta/unity/UCCatalogManagedClientSuite.scala [hunk 1]
```diff
@@ -55,22 +55,6 @@ class UCCatalogManagedClientSuite extends AnyFunSuite with UCCatalogManagedTestU
     new UCCatalogManagedClient(ucClient)
   }
 
-  /** Helper method with reasonable defaults */
-  private def loadSnapshot(
```
Comment (scottsand-db) [inline unity/src/test/scala/io/delta/unity/UCCatalogManagedClientSuite.scala:59]: moved to util trait

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java [hunk 2]
```diff
@@ -280,4 +341,18 @@ public CreateCheckpointIterator getCreateCheckpointIterator(Engine engine) {
   public Optional<Long> getLatestTransactionVersion(Engine engine, String applicationId) {
     return logReplay.getLatestTransactionIdentifier(engine, applicationId);
   }
+
+  private long getMaxPublishedDeltaVersionOrThrow() {
+    // The maxPublishedDeltaVersion is required for publishing to ensure published deltas are
+    // contiguous. The cases where it is unknown should be very rare (e.g. Kernel loaded a
+    // LogSegment consisting only of a checkpoint with no corresponding published delta).
+    // TODO: Kernel should LIST to authoritatively determine the maxPublishedDeltaVersion, or give
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:353]: We can do this later -- no a P0 for this PR

File: kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java [hunk 3]
```diff
@@ -101,4 +102,20 @@ public interface Snapshot {
    * @since 3.4.0
    */
   UpdateTableTransactionBuilder buildUpdateTableTransaction(String engineInfo, Operation operation);
+
```
Comment (allisonport-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java:105]: What's the current status/plan for exposing stats w.r.t. deciding whether to publish or not?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java [hunk 4]
```diff
@@ -186,6 +196,57 @@ public UpdateTableTransactionBuilder buildUpdateTableTransaction(
...
+
+    if (!allCatalogCommits.isEmpty()) {
+      if (isFileSystemBasedTable) {
+        throw new IllegalStateException( // This case should be impossible
+            "Cannot have catalog commits on a filesystem-managed table");
+      }
+
+      if (!committerSupportsPublishing) {
+        throw new UnsupportedOperationException( // This case should also be impossible
+            "Cannot publish: committer does not support publishing");
```
Comment (allisonport-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java]: is it possible to put committer class in the error message here as well? i.e. this could happen if a connector is mis-configuring things right?

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/LogSegment.java [hunk 5]
```diff
@@ -58,16 +60,20 @@ public static LogSegment createForNewTable(Path logPath, ParsedDeltaData parsedD
...
         logPath,
         0 /* version */,
         deltas,
         compactions,
         checkpoints,
-        deltaFile,
+        deltaFile /* deltaAtEndVersion */,
         Optional.empty() /* lastSeenChecksum */,
-        Optional.empty() /* maxPublishedDeltaVersion */);
+        maxPublishedDeltaVersion);
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/LogSegment.java:76]: Found a bug!

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java [hunk 6]
```diff
@@ -186,6 +196,57 @@ public UpdateTableTransactionBuilder buildUpdateTableTransaction(
...
 
+  @Override
+  public void publish(Engine engine) throws PublishFailedException {
+    final List<ParsedCatalogCommitData> allCatalogCommits = getLogSegment().getAllCatalogCommits();
+    final boolean isFileSystemBasedTable = !TableFeatures.isCatalogManagedSupported(protocol);
+    final boolean isCatalogCommitter = committer instanceof CatalogCommitter;
+
+    if (!allCatalogCommits.isEmpty()) {
+      if (isFileSystemBasedTable) {
+        throw new IllegalStateException( // This case should be impossible
```
Comment (scottsand-db) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java:207]: IMO not worth testing these edge/impossible cases -- code is trivial; we have tests for the positive cases;

File: unity/src/test/scala/io/delta/unity/InMemoryUCClient.scala [hunk 7]
```diff
@@ -91,6 +91,21 @@ object InMemoryUCClient {
...
+        throw new IllegalArgumentException(s"Version must be non-negative, but got: $version")
+      }
+
+      // Remove all commits with version <= the specified version
+      val indexToRemove = commits.lastIndexWhere(_.getVersion <= version)
+      if (indexToRemove >= 0) {
+        commits.remove(0, indexToRemove + 1)
+      }
+
+      // Note: We don't update maxRatifiedVersion as it represents the highest version
```
Comment (scottsand-db) [inline unity/src/test/scala/io/delta/unity/InMemoryUCClient.scala]: Delete -- cursor slop

File: unity/src/test/scala/io/delta/unity/UCE2ESuite.scala [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      val ucClient = new InMemoryUCClient("ucMetastoreId")
+      val ucCatalogManagedClient = new UCCatalogManagedClient(ucClient)
+
+      // Step 1: CREATE -- v0.json
+      val result0 = ucCatalogManagedClient
+        .buildCreateTableTransaction("ucTableId", tablePath, testSchema, "test-engine")
+        .build(engine)
+        .commit(engine, CloseableIterable.emptyIterable() /* dataActions */ )
+      val tableData0 = new TableData(-1, ArrayBuffer[Commit]())
+      ucClient.createTableIfNotExistsOrThrow("ucTableId", tableData0)
```
Comment (allisonport-db) [inline unity/src/test/scala/io/delta/unity/UCE2ESuite.scala:41]: Worth calling publish here too?

File: kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java (hunk 3 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java:105]: Will add that soon :) This week 👍

Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java:105]: I decided just to do this first as Kernel consumers could choose to start invoking this API immediately 👍

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java (hunk 4 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/SnapshotImpl.java]: Sg!

File: kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java (hunk 3 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java:105]: That's fine order-wise, I was just asking what the "that" is? haha, what exactly did we decide to add?

File: unity/src/test/scala/io/delta/unity/UCE2ESuite.scala (hunk 8 above)
Comment (scottsand-db) [reply unity/src/test/scala/io/delta/unity/UCE2ESuite.scala:41]: Just to be clear: Do you mean: Call publish here on the newly-created table? And are you aware that it was created via writing 000.json (commit 0 is file-system based for UC, not uuid)

Basically I'm just asking to confirm you want me to just double check that calling publish on `00.json` does nothing?

Sorry if that semed verbose lol -- IMO i'm fine not doing that? We have checks later for idempotency / calling publish on a fully published Snapshot?

File: kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java (hunk 3 above)
Comment (allisonport-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java:105]: is it just going to be `maxPublishedVersion` or something like `numUnpublishedCommits`

File: unity/src/test/scala/io/delta/unity/UCE2ESuite.scala (hunk 8 above)
Comment (allisonport-db) [reply unity/src/test/scala/io/delta/unity/UCE2ESuite.scala:41]: Yeah. I thought the 0 version case might be a slightly different edge-case. Saw some new table based code you fixed in the PR right?

File: kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java (hunk 3 above)
Comment (scottsand-db) [reply kernel/kernel-api/src/main/java/io/delta/kernel/Snapshot.java:105]: Something like: `numCommitsToPublish`

File: unity/src/test/scala/io/delta/unity/UCE2ESuite.scala (hunk 8 above)
Comment (scottsand-db) [reply unity/src/test/scala/io/delta/unity/UCE2ESuite.scala:41]: sg! will add 👍
